### PR TITLE
install/kubernetes: remove privileged from Cilium DaemonSet

### PR DESCRIPTION
**Pull request by aanm**
We are already setting all the capabilities that Cilium requires so we
don't need to set privileged as true.

Signed-off-by: André Martins <andre@cilium.io>

```release-note
Remove privileged mode in Cilium's DaemonSet
```

**Note for reviewers**: all tests have passed on PR https://github.com/cilium/cilium/pull/19449 which includes these changes.

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-missed-k8s

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml [hunk 1]
```diff
@@ -372,7 +372,6 @@ spec:
           capabilities:
             add:
             - NET_ADMIN
-          privileged: true
```
Comment (sayboras) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: maybe we can have a quick look for this secion https://github.com/cilium/cilium/blame/master/Documentation/operations/system_requirements.rst#L387-L396 and check if any change is required.

**Review by sayboras — APPROVED** [review]
LGTM 💯 , one minor comment on docs.

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml (hunk 1 above)
Comment (aanm) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: Good spotting, yes we are looking into the capabilities that Cilium needs. @borkmann the docs says `CAP_SYS_ADMIN` is needed but Cilium only have `NET_ADMIN`, does `NET_ADMIN` also have `SYS_ADMIN`?

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
retest-net-next

**Comment by aanm** [comment]
retest-net-next

**Comment by aanm** [comment]
retest-net-next

File: bpf/init.sh [hunk 2]
```diff
@@ -14,30 +14,32 @@
...
 
-LIB=$1
-RUNDIR=$2
-IP4_HOST=$3
-IP6_HOST=$4
-MODE=$5
+LIB=${1}
+RUNDIR=${2}
+PROCSYSNETDIR=${3}
+SYSDIR=${4}
```
Comment (kaworu) [inline bpf/init.sh]: nit: Since we're only using `${SYSDIR}/class/net`, the "asymmetry" with `${PROCSYSNETDIR}` looks odd to me. In other words, either one of:

* `${PROCDIR}` for `/proc` and `${SYSDIR}` for `/sys`
* `${PROCSYSNETDIR}` for `/proc/sys/net` and `${SYSCLASSNETDIR}` for `/sys/class/net`

would "feel cleaner", does that make sense?

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
retest-runtime

**Comment by aanm** [comment]
retest-4.9

**Comment by aanm** [comment]
retest-4.9

**Comment by aanm** [comment]
retest-4.9

**Comment by aanm** [comment]
retest-4.9

**Comment by aanm** [comment]
test-me-please

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml [hunk 3]
```diff
@@ -230,13 +231,51 @@ spec:
...
+              # Check if monitor breaks, we might need:
+              # 'access privileged perf event information;'
+              # Check if we still receive fsnotify events because of:
+              # 'call fanotify_init(2);'
+              - SYS_ADMIN
+              # Could be an alternative for the SYS_ADMIN for the RLIMIT_NPROC
+              - SYS_RESOURCE
+              # Re-add if endpoint create fails or if the log driver is using syslog
+              #- SYSLOG
+              # Requires kernel 5.8 and container runtime X
```
Comment (errordeveloper) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: What's X?

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml [hunk 4]
```diff
@@ -230,13 +231,51 @@ spec:
...
+              #- PERFMON
+              #- BPF
         volumeMounts:
+        - mountPath: /host/proc/sys/net
+          name: host-proc-sys-net
+        - mountPath: /host/proc/sys/kernel
+          name: host-proc-sys-kernel
 {{- /* CRI-O already mounts the BPF filesystem */ -}}
 {{- if not (eq .Values.containerRuntime.integration "crio") }}
-        - mountPath: /sys/fs/bpf
```
Comment (errordeveloper) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: This looks like a cosmetic change, but it certainly deserves a separate commit!

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml [hunk 5]
```diff
@@ -230,13 +231,51 @@ spec:
...
+              # Could be an alternative for the SYS_ADMIN for the RLIMIT_NPROC
+              - SYS_RESOURCE
+              # Re-add if endpoint create fails or if the log driver is using syslog
+              #- SYSLOG
+              # Requires kernel 5.8 and container runtime X
+              # if enabled user can remove SYS_ADMIN
+              #- PERFMON
+              #- BPF
         volumeMounts:
+        - mountPath: /host/proc/sys/net
```
Comment (errordeveloper) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: So with `privileged: true` the pod was getting `/proc/sys/net` and `/proc/sys/kernel` automatically? I'd say that it needs to be stated more explicitly in the commit message, as it looks like a significant change and not at all obvious. I would even consider putting it in a separate commit also, which can go before the change to `securityContext`.

**Review by errordeveloper — CHANGES_REQUESTED** [review]
This is really cool, and I think this is complex enough to be broken into 3 commits like this:

1) bind-mount `/proc/sys` dirs explicitly from the host (prepare to remove privileged mode)
2) use least privileges for agent container
3) change path to BPF filesystem for consistency

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-runtime

**Comment by aanm** [comment]
test-1.19-5.4

**Review by michi-covalent — APPROVED** [review]

**Comment by aanm** [comment]
/test-1.16-4.9

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 6]
```diff
@@ -233,13 +239,53 @@ spec:
         {{- end }}
         {{- end }}
         securityContext:
+          {{- if .Values.securityContext.privileged }}
           privileged: true
+          {{- else }}
```
Comment (JAORMX) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:246]: if the `privileged` flag is not set, you can disable make the container "unconfined" for SELinux (which would be needed here), by setting the following:

```
securityContext:
  seLinuxOptions:
    level: 's0'
    type: 'spc_t'
```

The level might not have been strictly needed here, but it's good to add just in case.

**Review by JAORMX — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 7]
```diff
@@ -825,6 +825,7 @@ data:
   enable-bgp-control-plane: "false"
 {{- end }}
 
+  procfs: {{ .Values.procfs | quote }}
```
Comment (joestringer) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: I would have expected that the user may want to choose where the _host_ procfs is, so that we can map that into the container. I don't understand why we would let the user configure the path _inside_ the container, since there's only one correct configuration parameter for this value (otherwise it's incompatible with the hardcoded changes in the `DaemonSet`).

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 8]
```diff
@@ -233,13 +241,59 @@ spec:
...
+          seLinuxOptions:
+            level: 's0'
+            # Running with spc_t since we have removed the privileged mode.
+            # Users can change it to a different type as long as they have the
+            # type available on the system.
+            type: 'spc_t'
```
Comment (joestringer) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:252]: Does this mean that privileged=false requires SELinux?

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 9]
```diff
@@ -362,7 +416,45 @@ spec:
...
+        args:
+        - 'mount | grep "/sys/fs/bpf type bpf" || mount -t bpf bpf /sys/fs/bpf'
+        command:
+        - /bin/bash
+        - -c
+        - --
+        securityContext:
+          privileged: true
+      {{- /* CRI-O already mounts the BPF filesystem */ -}}
+      {{- if not (eq .Values.containerRuntime.integration "crio") }}
```
Comment (joestringer) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:459]: This part seems like it should be above the whole init container that's defined here..?

File: pkg/sysctl/sysctl.go [hunk 10]
```diff
@@ -54,7 +60,31 @@ func parameterPath(name string) (string, error) {
...
-	return filepath.Join(append([]string{prefixDir}, elems...)...), nil
+	return filepath.Join(append([]string{GetProcfs(), "sys"}, elems...)...), nil
+}
+
+// SetProcfs sets path for the root's /proc. Setting this value with a different
+// default value and executing it after GetProcfs causes panic.
+func SetProcfs(path string) {
+	if path == procFsDefault {
+		return
+	}
```
Comment (kaworu) [inline pkg/sysctl/sysctl.go]: Can you elaborate the rational behind this codepath? I find it confusing that `SetProcfs(procFsDefault)` is consistently a no-op:
```golang
SetProcfs("/foo")
SetProcfs(procFsDefault)
p := GetProcfs() // here p == "/foo"
SetProcfs(procFsDefault) // no panic
SetProcfs("/foo") // panic although could be idem-potent
```

**Review by kaworu — COMMENTED** [review]
One small question. I'm deferring review here as I'm lacking context.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml (hunk 8 above)
Comment (aanm) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:252]: No no, it only means that if SELinux is available / enabled then we should be running the container with level `s0` and type `spc_t`

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml (hunk 9 above)
Comment (aanm) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:459]: Good point, however I don't think it hurts running this init container even if cri-o already mounts the BPF filesystem. Let's assume worst case where cri-o mounts the directory inside the container but the bpf fs is not mounted.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml (hunk 7 above)
Comment (aanm) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: Oh, very good catch, it does not make sense to expose this option to the user. It should be this logic instead:
```
{{- if not .Values.securityContext.privileged }}
  procfs: "/host/proc"
{{- end }}
```

File: pkg/sysctl/sysctl.go (hunk 10 above)
Comment (aanm) [reply pkg/sysctl/sysctl.go]: Yeah I don't see any rational behind it, this code was written over 1y ago. Removed.

File: Documentation/helm-values.rst [hunk 11]
```diff
@@ -1381,6 +1381,10 @@
      - The priority class to use for cilium-agent.
      - string
      - ``""``
+   * - procfs
+     - 
+     - string
+     - ``"/host/proc"``
```
Comment (joestringer) [inline Documentation/helm-values.rst]: Has this file been regenerated? I think you removed this option, and it shouldn't need to be exposed to the user any more...?

**Review by joestringer — APPROVED** [review]
Just one nit, seems like the helm-values docs page didn't get updated after the last change. Other than that, LGTM

**Review by qmonnet — APPROVED** [review]

**Review by joestringer — APPROVED** [review]

**Review by christarazi — APPROVED** [review]

File: pkg/sysctl/sysctl.go [hunk 12]
```diff
@@ -54,7 +60,28 @@ func parameterPath(name string) (string, error) {
 			return "", ErrInvalidSysctlParameter(name)
 		}
 	}
-	return filepath.Join(append([]string{prefixDir}, elems...)...), nil
+	return filepath.Join(append([]string{GetProcfs(), "sys"}, elems...)...), nil
+}
+
+// SetProcfs sets path for the root's /proc. Setting this value with a different
+// default value and executing it after GetProcfs causes panic.
```
Comment (kaworu) [inline pkg/sysctl/sysctl.go]: This comment is still referring to the previous behavior
```suggestion
// SetProcfs sets path for the root's /proc. Calling it after GetProcfs causes panic.
```

**Comment by aanm** [comment]
The CI has passed in https://github.com/cilium/cilium/pull/19449 although AKS test is failing on master so it's not a regression introduced by these changes. Merging...
